python_tests/test_cached_extra_study_property.py

```diff
@@ -0,0 +1,177 @@
+                "x0": 0.5,
+                "x1": 0.5,
+            },
+            {
+                "x0": 0.5,
+            },
+            {
+                "x0": 0.5,
+                "x1": 0.5,
+            },
+        ]
+        trials = [
+            create_trial(state=TrialState.COMPLETE, value=0, distributions=d, params=p)
+            for d, p in zip(distributions, params)
+        ]
+
+        cached_extra_study_property = _CachedExtraStudyProperty()
+        cached_extra_study_property.update(trials)
+
+        self.assertEqual(len(cached_extra_study_property.intersection), 0)
+        self.assertEqual(len(cached_extra_study_property.union), 3)
+
+    def test_no_intermediate_value(self) -> None:
+        intermediate_values = [
+            {},
+            {},
+        ]
+        trials = [
+            create_trial(
+                state=TrialState.COMPLETE,
+                value=0,
+                distributions={"x0": UniformDistribution(low=0, high=10)},
+                intermediate_values=iv,
+                params={"x0": 0.5},
+            )
+            for iv in intermediate_values
+        ]
+        cached_extra_study_property = _CachedExtraStudyProperty()
+        cached_extra_study_property.update(trials)
+        self.assertFalse(cached_extra_study_property.has_intermediate_values)
+
+    def test_some_trials_has_no_intermediate_value(self) -> None:
+        intermediate_values = [
+            {0: 0.3, 1: 1.2},
+            {},
+            {0: 0.3, 1: 1.2},
+        ]
+        trials = [
+            create_trial(
+                state=TrialState.COMPLETE,
+                value=0,
+                distributions={"x0": UniformDistribution(low=0, high=10)},
+                intermediate_values=iv,
+                params={"x0": 0.5},
+            )
+            for iv in intermediate_values
+        ]
+        cached_extra_study_property = _CachedExtraStudyProperty()
+        cached_extra_study_property.update(trials)
+        self.assertTrue(cached_extra_study_property.has_intermediate_values)
+
+    def test_all_trials_has_intermediate_value(self) -> None:
+        intermediate_values = [{0: 0.3, 1: 1.2}, {0: 0.3, 1: 1.2}]
+        trials = [
+            create_trial(
+                state=TrialState.COMPLETE,
+                value=0,
+                distributions={"x0": UniformDistribution(low=0, high=10)},
+                intermediate_values=iv,
+                params={"x0": 0.5},
+            )
+            for iv in intermediate_values
+        ]
+        cached_extra_study_property = _CachedExtraStudyProperty()
+        cached_extra_study_property.update(trials)
+        self.assertTrue(cached_extra_study_property.has_intermediate_values)
+
+    def test_no_trials(self) -> None:
+        trials = []
+        cached_extra_study_property = _CachedExtraStudyProperty()
+        cached_extra_study_property.update(trials)
+        self.assertFalse(cached_extra_study_property.has_intermediate_values)
```

Comment:
How about splitting these test cases into another TestCase class? This test class looks somewhat large. These logics are defined in the same class (`_CachedExtraStudyProperty`) though.